### ISSUE TITLE
max_ext, min_ext, mean_ext, median_ext

### ISSUE THREAD

**Issue by tabularelf** [Alphish/gm-community-toolbox#17]
I've found that while `max()`, `min()`, `mean()` and `median()` are very nice, it would be a lot more nicer being able to pass in an array of values instead of passing a predefined set of arguments. So I'd like to suggest a few new math functions to compliment the gm community toolbox. As others may find it just as useful, if not more useful than I do.

I mainly use `max_ext` and `min_ext` in my own code, but I'll be including `mean_ext` and `median_ext` to make them complete.
`script_execute_ext()` has been for a long while supported built in functions, so this makes this sort of implementation super easy, while allowing as many arguments as we'd like. I've considered going with `*_array`, but I think with the pattern of how GML goes by it's naming convention, `*_ext` is more fitting. The descriptions themselves could be expanded upon admittedly, but the idea is there. The implementation itself doesn't really take up too much either.

```gml
/// @func max_ext(array)
/// @param {Array} array
/// @desc Returns the max number value from the array.
/// @returns {Real}
function max_ext(array) {
    return script_execute_ext(max, array);
} 
```

```gml
/// @func min_ext(array)
/// @param {Array} array
/// @desc Returns the min number value from the array.
/// @returns {Real}
function min_ext(array) {
    return script_execute_ext(min, array);
} 
```

```gml
/// @func mean_ext(array)
/// @param {Array} array
/// @desc Adds up all of the numbers from the array, divides them by their array length, and returns that number.
/// @returns {Real}
function mean_ext(array) {
    return script_execute_ext(mean, array);
} 
```

```gml
/// @func median_ext(array)
/// @param {Array} array
/// @desc Returns the middle most number value from the array. 
/// @returns {Real}
function median_ext(array) {
    return script_execute_ext(median, array);
} 
```

**Comment by tabularelf** [comment]
Accidentally preposted this while I was writing it up... Now that I've finished writing it up 😆

**Comment by Alphish** [comment]
Another option would be not `min_array` etc., but rather `array_min`, `array_max`, `array_mean` and `array_median`.

Then since we're doing maths, we might as well consider adding `array_sum`, though this one doesn't have the straightforward implementation of `script_execute_ext(...)`, as far as I can tell.

We might even go as far as doing `array_sum_by`, `array_min_by`, `array_max_by`, `array_mean_by` and `array_median_by` (with values to sum/compare produced from item-derived values, rather than items themselves), though this might be considered separately. Guess we might focus on the array-based variants of `min`/`max`/etc. for now.

**Comment by tabularelf** [comment]
I like `array_min`, I just didn't think it was fitting since that could mean that you're returning the minimal value from the array from either a number or by a function for comparing against a specific value... Which would be pretty nice to have. 

But if we were wanting to have a `sum()` function and `array_sum()`, we would need to make those for sure. Which... We can easily do. And making it work with `script_execute_ext()` is super easy. Anything that can take in multiple arguments, will work.

```gml
/// @func sum(...)
/// @desc adds up all of the arguments and returns the total sum.
/// @return {Real}
function sum() {
      var i = 0; 
      var sum = 0;
      repeat(argument_count) {
          sum += argument[i];
          ++i;
      }
      return sum;
}

/// @func sum(...)
/// @param {Array} array
/// @desc adds up all of the values from the array and returns the total sum.
/// @return {Real}
function array_sum(array) {
      return script_execute_ext(sum, array);
}
```

**Comment by tabularelf** [comment]
Though in that case, it might be better to just rewrite both into one and support arrays only.

**Comment by Alphish** [comment]
Alright then, if no one opposes, I'd accept implementing the following functions:
- `array_min(array)`
- `array_max(array)`
- `array_mean(array)`
- `array_median(array)`
- `array_sum(array)` (without a separate `sum` function)

Some thing to clarify:
1. How do we handle 0-length arrays? Do we return 0, `undefined`, something else? Generally, the `script_execute_ext(...)` thing seems to return 0 for all these math functions, so it might be fine.
2. I've found that if there are two middle elements (because the number of array items is even), then GML returns the higher of them instead of the average of them. The question is, do we want to have mathematically correct median, or have "GML median but for array". Going by the familiarity principle (and also implementation simplicity), I'm inclined towards the latter.

If no-one disagrees with the functions lineup above and we settle these questions, it'll be ready for implementation.

**Comment by tabularelf** [comment]
1. For 0-length arrays, I'd be more inclined to throw an error, but even the normal math functions return `0`... So it'd be in a sense correcting an oddity for the default math functions if we were to do anything about it. Though that's probably something worth asking YYG for opinion on. If an empty function like `max()` is meant to return anything at all, or should it realistically error out, or return some other kind of value like `NaN`?

2. Looking over at the documentation and my own testing, median returns the lowest of the two even values. So I'm perplexed that you got higher. But I would prefer going with what GM goes by, as these are meant to be more close to existing GM-like functionality. No matter how *odd* you and I may feel about them not being an average between the two. But it does make some sense. Wouldn't want to return a value that just doesn't exist in your array. ^^

**Comment by Alphish** [comment]
Just for the record, here are some tests of mine:
```gml
var _max_none = script_execute_ext(max, []);
var _min_none = script_execute_ext(min, []);
var _mean_none = script_execute_ext(mean, []);
var _median_none = script_execute_ext(median, []);

show_debug_message($"Max none: {_max_none}");
show_debug_message($"Min none: {_min_none}");
show_debug_message($"Mean none: {_mean_none}");
show_debug_message($"Median none: {_median_none}");

var _median_some = script_execute_ext(median, [1, 10]);
show_debug_message($"Median some: {_median_some}");
```

And here is what I got on 2023.4 monthly:
```
Max none: 0
Min none: 0
Mean none: 0
Median none: 0
Median some: 10
```

In case of 0-length arrays, I'd be inclined towards returning NaN rather than throwing an error, based on my general preference of returning a non-result when possible (as opposed to throwing an error). Or we might decide not to correct the behaviour of corresponding GM functions, and return 0 like the underlying implementations do.

The latter would be more in-line with the familiarity design guideline, except documentation doesn't really specify `max` will return 0 for no arguments, either (who in their right mind would want to directly use max with 0 arguments, anyway). ^^'

Considering GM manual mentions median returning the lower value of two middle ones, I'd be fine with accepting that outcome... except as the print-out above shows, even for trivial case of two values it picks the larger one. So I guess it would count as a bug? It might be a good opportunity to contribute to the GM tests, I guess...

**Comment by tabularelf** [comment]
If we are wanting to keep it exactly as how GM implemented it, then it'd make sense to go with the exact cases.
I did also validate with `median(1, 10)` by itself just in case it wasn't some kind of odd edge case, and it does seem that `median` in general has a weird bug with it returning `10` instead of `1`. Might be some kind of by product?

Actually just testing it with other number combinations, I am *specifically* only getting the higher value. Guessing my own testing was more of a fluke of itself. 😅

In any case, that should be reported. 

As for empty arrays returning `0`? I'd be more imposed to send that to YYG and actually see about them looking to either add towards the documentation or correcting how the function *should* work. Outside of that... I guess the main thing about us returning something like `NaN` in our own case, is provided that we do document that it does for empty arrays, we should be okay. And we won't run into any potential future problems with the default math functions. However, it does mean that we would be deviating from what GM would do. But that's not necessarily a bad thing either...

**Comment by tabularelf** [comment]
TLDR: I'm okay with accepting `NaN` provided that we include that within the documentation.

**Comment by Alphish** [comment]
Just for the record, I checked the unit tests for GM, and they check for both the min/max/mean/median functions returning 0 for no arguments, and also for median returning the larger of middle values. The tests for the latter can be found here: https://github.com/YoYoGames/GM-TestFramework/blob/dd30b6fbe81d05f8f27b1b03ad2b95cfe7d59166/projects/xUnit/scripts/BasicMathTestSuite/BasicMathTestSuite.gml#L2936

So I'm inclined towards using the GM's tested behaviour, and having people check the array length of 0 themselves and handle it on their own. The JSDoc should include a mention of returning 0 for empty arrays, and returning the larger of two middle values as well.

**Comment by tabularelf** [comment]
Interesting! Well, I guess that answers that. 😆

**Comment by Alphish** [comment]
Since there are no further complaints or questions, the following functions are accepted for implementation:
- `array_min(array)`
- `array_max(array)`
- `array_mean(array)`
- `array_median(array)`
- `array_sum(array)`

The functions should go into `utils_Array` script, next to `utils_Math` and `utils_Misc`.

These functions will return 0 when given an empty array, and `array_median` will return the larger of two middle values, given an even non-zero number of arguments.

The tests should check regular use scenarios, passing in an empty array, and tests for `array_median` should separately check the scenario with odd and even number of arguments.

The demo might include some randomised array, with a varying number of arguments, each being an integer e.g. between -100 and 100. There might also be a button for re-randomising the array as well. The results of all the functions above would be displayed for the current array.

As usual, if someone decides to start implementation work, please write about it here so that multiple people won't end up duplicating each other's work.

**Comment by Mercerenies** [comment]
`script_execute_ext` makes me instinctively uncomfortable. Game Maker used to have a 16-argument limit on function calls, so would this limit us to arrays of 16 elements, or does that limit not apply to `script_execute_ext` (I don't have GM in front of me to test right now)?

And even if it is allowed, in my experience in other languages, converting arrays into parameter lists causes performance bottlenecks when the array is large. I'd rather see the implementation of each of these in terms of a `for` loop, rather than dumping into a parameter list.

**Comment by tabularelf** [comment]
> script_execute_ext makes me instinctively uncomfortable. Game Maker used to have a 16-argument limit on function calls, so would this limit us to arrays of 16 elements, or does that limit not apply to script_execute_ext (I don't have GM in front of me to test right now)?

This doesn't apply to `script_execute_ext`. It can take as many arguments as you want. Also, `max()`, `median()`, `mean()` and `min()` can take as many arguments as you wish. This works just fine. 
```gml
max(10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150, 160, 170, 180, 190, 200);
```
There's only a limit of 16 arguments if the function isn't designed to allow for infinite arguments.

> And even if it is allowed, in my experience in other languages, converting arrays into parameter lists causes performance bottlenecks when the array is large. I'd rather see the implementation of each of these in terms of a for loop, rather than dumping into a parameter list.

There's a few reasons why I disagree on this.
1. Letting GameMaker take the wheel on anything is a performance boost, much faster than a for loop or a repeat loop.
2. While I don't fully understand the underlying implementation, GameMa
ker does handle a lot of footwork much better than passing in an array or multiple arguments in. 

In fact, I can even prove that it's faster than anything else. For the record, I am using <https://github.com/DragoniteSpam/GMBenchmark> to conduct my benchmarking on this.

First screenshot here just covers an array element with the exact values as from above.
<img width="450" alt="Runner_2023-07-06_01-43-34" src="https://github.com/Alphish/gm-community-toolbox/assets/26135221/39dc4f2e-69a0-4fa1-9541-f50cdafe0582">

And in the second screenshot, I made a function called `max_alternate()`, which takes in an array and does it manually.
```gml
function max_alternate(_array) {
	var _num = -infinity;
	var _i = 0;
	repeat(array_length(_array)) {
		if (_array[_i] > _num) {
			_num = _array[_i];	
		}
		++_i;
	}
	return _num;
}
```
 Using `repeat()` as the fastest form of looping. And you know what the results on that was?
<img width="450" alt="Runner_2023-07-06_01-48-31" src="https://github.com/Alphish/gm-community-toolbox/assets/26135221/ca8a5a26-b326-4933-9c74-dfc7ae1987f2">
**Slow!!!** I also tested this in YYC, and while `max_alternate()` does improve a fair bit (and could probably squeeze a bit more with `gml_pragma("forceinline")`, `max()` and `script_execute_ext()` are just about equally on par. And this is keeping in mind, that they're being fed the same array, preallocated *once* prior to the benchmark test.
<img width="450" alt="GMBenchmark_2023-07-06_01-53-44" src="https://github.com/Alphish/gm-community-toolbox/assets/26135221/4ffe7c6c-1a97-4cf1-ac4d-e2257c7a2d1f">

```gml
new Benchmark("Max calls vs script_execute_ext", [
        new TestCase("max()", function(iterations) {
            repeat (iterations) {
                var result = max(10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150, 160, 170, 180, 190, 200);
            }
        }),
		 new TestCase("max_alternative()", function(iterations) {
			var _array = [10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150, 160, 170, 180, 190, 200];
            repeat (iterations) {
                var result = max_alternate(_array);
            }
        }),
        new TestCase("script_execute_ext(max, _array)", function(iterations) {
			var _array = [10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150, 160, 170, 180, 190, 200];
            repeat (iterations) {
                var result = script_execute_ext(max, _array);
            }
        })
    ])
    ```

**Comment by tabularelf** [comment]
Hopefully this gives some further insight on the matter at hand, and why I chose to go with this route. ^^

**Comment by Mercerenies** [comment]
@tabularelf Yeah, that makes a ton of sense. I can't argue with results. Thank you for the detailed rebuttal.

So I'm guessing GM compiles variable-argument functions down to a function that takes a single array argument, and then the actual engine code is able to iterate over that array faster than we can. I was imagining that GM did what C and Lisp do, which is dump all of the arguments onto the call stack individually, but based on those benchmarks I have to conclude that I was wrong about that.

**Comment by Alphish** [comment]
The implementation and tests are already merged, only the demo remains.

**Comment by Alphish** [comment]
The demo has been merged in Pull Request #31 as well!
